tools/shared.py

```diff
@@ -1292,13 +1293,12 @@ def __getattr__(self, attr):
 
     def __setattr__(self, attr, value):
       if attr not in self.attrs:
-        import difflib
-        logging.warning('''Assigning a non-existent settings attribute "%s"''' % attr)
+        logging.error('Assigning a non-existent settings attribute "%s"' % attr)
         suggestions = ', '.join(difflib.get_close_matches(attr, list(self.attrs.keys())))
         if suggestions:
-          logging.warning(''' - did you mean one of %s?''' % suggestions)
-        logging.warning(''' - perhaps a typo in emcc's  -s X=Y  notation?''')
-        logging.warning(''' - (see src/settings.js for valid values)''')
+          logging.error(' - did you mean one of %s?' % suggestions)
+        logging.error(" - perhaps a typo in emcc's  -s X=Y  notation?")
+        exit_with_error(' - (see src/settings.js for valid values)')
```

Comment:
This looks like this line get treated specially by `exit_with_error`. Might be slightly simpler to leave the consistency of logging with `logging.error` calls, and separate out the `sys.exit(1)`, though that slightly duplicates the exit-with-error logic, so who knows.